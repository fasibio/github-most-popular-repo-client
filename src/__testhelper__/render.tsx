import { ReactElement } from 'react';
import {
  render as realRender,
  fireEvent,
  RenderOptions,
  RenderResult,
} from '@testing-library/react';

import { Grommet } from 'grommet';

export const render = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
): RenderResult => {
  return realRender(<Grommet>{ui}</Grommet>, options);
};
