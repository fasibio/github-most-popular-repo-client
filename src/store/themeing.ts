import { customTheme as aruba } from '../themes/aruba';
import { customTheme as hpe } from '../themes/hpe';
import { customTheme as materialDark } from '../themes/materialDark';
import { customTheme as materialLight } from '../themes/materialLight';
import { customTheme as metroTheme } from '../themes/metro';
import { customTheme as purpleTheme } from '../themes/purpleTheme';
import create from 'zustand';

type Theme = any;

interface ThemeObj {
  name: string;
  theme: Theme;
}

interface ThemeModeState {
  dark: boolean;
  setMode: (dark: boolean) => void;
  getMainBackground: () => string;
}

export const useThemeMode = create<ThemeModeState>((set, get) => ({
  dark: true,
  setMode: (dark) => {
    set({ dark });
  },
  getMainBackground: () => (get().dark ? 'dark-1' : 'light-1'),
}));

interface ThemeingState {
  themes: ThemeObj[];
  selectedTheme: ThemeObj;
  setSelectedTheme: (t: ThemeObj) => void;
}

export const useThemeing = create<ThemeingState>((set) => ({
  themes: [
    {
      name: 'Pink',
      theme: purpleTheme,
    },
    {
      name: 'Ligth',
      theme: materialLight,
    },
    {
      name: 'Dark',
      theme: materialDark,
    },
    {
      name: 'Metro',
      theme: metroTheme,
    },
    {
      name: 'Aruba',
      theme: aruba,
    },
    {
      name: 'Hpe',
      theme: hpe,
    },
  ],
  selectedTheme: {
    name: 'pink',
    theme: purpleTheme,
  },
  setSelectedTheme: (theme) => set({ selectedTheme: theme }),
}));