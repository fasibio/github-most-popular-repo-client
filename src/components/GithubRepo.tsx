import {
  Anchor,
  Box,
  Button,
  Card,
  CardBody,
  CardHeader,
  Tab,
  Tabs,
  Text,
} from 'grommet';
import { Book, Bookmark, Star } from 'grommet-icons';
import { useTranslation } from 'react-i18next';
import { Repo } from '../store/github-data';
export interface Props {
  repo: Repo;
  isBookedMarked: boolean;
  onBookmarkClicked: (repo: Repo) => void;
}

export const GithubRepo = ({ repo, onBookmarkClicked, isBookedMarked }: Props) => {
  const { t } = useTranslation();
  return (
    <Card width={{ max: 'large' }}>
      <CardHeader background={'background-front'}>
        <Box margin="small" direction="row" justify={'between'} flex={true}>
          <Box direction="row" gap={'small'}>
            <Book />
            <Anchor
              target={'_blank'}
              href={repo.owner.html_url}
              size="medium"
              label={repo.owner.login}
            />
            <Text size="medium">/</Text>
            <Anchor
              target={'_blank'}
              href={repo.html_url}
              size="medium"
              label={repo.name}
            />
          </Box>

          <Box direction="column" align="center" gap="small">
            <Button
              data-testid="bookmarkbutton"
              alignSelf="end"
              title={t('githubrepo.btnAddToFavoriteToolTip') as string}
              icon={<Bookmark color={isBookedMarked ? 'focus' : undefined} />}
              onClick={() => {
                onBookmarkClicked(repo);
              }}
            />
            <Box direction="row" align="center" gap="xxsmall">
              <Star />
              <Text>{repo.stargazers_count}</Text>
            </Box>
          </Box>
        </Box>
      </CardHeader>
      <CardBody background={'background-back'}>
        <Tabs>
          <Tab title={t('githubrepo.tabs.0.title')}>
            <Box margin={'small'}>
              <Text>{repo.description ?? t('githubrepo.tabs.0.noDescription')}</Text>
            </Box>
          </Tab>
          <Tab
            title={`${t('githubrepo.tabs.1.title')} (${repo.open_issues_count})`}
            onClick={() => {
              window.open(`${repo.html_url}/issues`);
            }}
          />
        </Tabs>
      </CardBody>
    </Card>
  );
};
