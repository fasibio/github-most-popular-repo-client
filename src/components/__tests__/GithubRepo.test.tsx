import { GithubRepo } from '../GithubRepo';
import { render } from '../../__testhelper__/render';
import { Owner, Repo } from '../../store/github-data';
import type { PartialDeep } from 'type-fest';

describe('tests githubrepo card component', () => {
  const renderTests: {
    name: string;
    input: {
      repo: PartialDeep<Repo>;
      isBookedMarked: boolean;
    };
  }[] = [
    {
      name: 'fullsetTest',
      input: {
        repo: {
          id: 12344,
          html_url: 'repo html_url',
          name: 'repo name',
          stargazers_count: 160,
          description: 'test Description',
          open_issues_count: 2,
          owner: {
            html_url: 'owner html_url',
            login: 'login',
          },
        },
        isBookedMarked: true,
      },
    },
    {
      name: 'fullsetTest not bookmarked',
      input: {
        repo: {
          id: 12344,
          html_url: 'repo html_url',
          name: 'repo name',
          stargazers_count: 160,
          description: 'test Description',
          open_issues_count: 2,
          owner: {
            html_url: 'owner html_url',
            login: 'login',
          },
        },
        isBookedMarked: false,
      },
    },
    {
      name: 'fullsetTest no description',
      input: {
        repo: {
          id: 12344,
          html_url: 'repo html_url',
          name: 'repo name',
          stargazers_count: 160,
          description: undefined,
          open_issues_count: 2,
          owner: {
            html_url: 'owner html_url',
            login: 'login',
          },
        },
        isBookedMarked: true,
      },
    },
  ];
  renderTests.forEach((test) => {
    it(`tests render componen ==> ${test.name}`, () => {
      const component = render(
        <GithubRepo
          onBookmarkClicked={() => {}}
          isBookedMarked={test.input.isBookedMarked}
          repo={test.input.repo as Repo}
        />,
      );
      expect(component.container).toMatchSnapshot();
    });
  });

  it('tests click on Bookmarkbutton will be triggered', async () => {
    const mockData = {
      id: 12344,
      html_url: 'repo html_url',
      name: 'repo name',
      stargazers_count: 160,
      description: 'test Description',
      open_issues_count: 2,
      owner: {
        html_url: 'owner html_url',
        login: 'login',
      },
    };
    const onBookmarkClickedMock = jest.fn();
    const component = render(
      <GithubRepo
        onBookmarkClicked={onBookmarkClickedMock}
        isBookedMarked={true}
        repo={mockData as Repo}
      />,
    );
    const button = await component.findByTestId('bookmarkbutton');
    button.click();
    expect(onBookmarkClickedMock).toBeCalledWith(mockData);
  });
});
