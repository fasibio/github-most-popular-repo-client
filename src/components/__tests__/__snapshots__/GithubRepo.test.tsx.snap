// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests githubrepo card component tests render componen ==> fullsetTest 1`] = `
<div>
  <div
    class="StyledGrommet-sc-19lkkz7-0 ioQEen"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 hECOWc"
    >
      <header
        class="StyledBox-sc-13pk1d4-0 cUsmfa"
      >
        <div
          class="StyledBox-sc-13pk1d4-0 hyJwyv"
        >
          <div
            class="StyledBox-sc-13pk1d4-0 euWtCA"
          >
            <svg
              aria-label="Book"
              class="StyledIcon-sc-ofa7kd-0 jQvEgu"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 1v10l3-2 3 2V1M5.5 18a2.5 2.5 0 1 0 0 5H22M3 20.5v-17A2.5 2.5 0 0 1 5.5 1H21v17.007H5.492M20.5 18a2.5 2.5 0 1 0 0 5"
                fill="none"
                stroke="#000"
                stroke-width="2"
              />
            </svg>
            <div
              class="StyledBox__StyledBoxGap-sc-13pk1d4-1 ilMzAg"
            />
            <a
              class="StyledAnchor-sc-1rp7lwl-0 dSRcaq"
              href="owner html_url"
              target="_blank"
            >
              login
            </a>
            <div
              class="StyledBox__StyledBoxGap-sc-13pk1d4-1 ilMzAg"
            />
            <span
              class="StyledText-sc-1sadyjn-0 jbJJSg"
            >
              /
            </span>
            <div
              class="StyledBox__StyledBoxGap-sc-13pk1d4-1 ilMzAg"
            />
            <a
              class="StyledAnchor-sc-1rp7lwl-0 dSRcaq"
              href="repo html_url"
              target="_blank"
            >
              repo name
            </a>
          </div>
          <div
            class="StyledBox-sc-13pk1d4-0 byddCk"
          >
            <button
              class="StyledButton-sc-323bzc-0 bXHvkf"
              data-testid="bookmarkbutton"
              title="githubrepo.btnAddToFavoriteToolTip"
              type="button"
            >
              <svg
                aria-label="Bookmark"
                class="StyledIcon-sc-ofa7kd-0 iUTzQu"
                viewBox="0 0 24 24"
              >
                <path
                  d="M5 1v21l7-5 7 5V1z"
                  fill="none"
                  stroke="#000"
                  stroke-width="2"
                />
              </svg>
            </button>
            <div
              class="StyledBox__StyledBoxGap-sc-13pk1d4-1 drdMXk"
            />
            <div
              class="StyledBox-sc-13pk1d4-0 dEmnTc"
            >
              <svg
                aria-label="Star"
                class="StyledIcon-sc-ofa7kd-0 jQvEgu"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 16.667 5 22l3-8-6-4.5h7.5L12 2l2.5 7.5H22L16 14l3 8z"
                  fill="#FFC95E"
                  fill-rule="evenodd"
                />
              </svg>
              <div
                class="StyledBox__StyledBoxGap-sc-13pk1d4-1 cuTpwc"
              />
              <span
                class="StyledText-sc-1sadyjn-0 jbJJSg"
              >
                160
              </span>
            </div>
          </div>
        </div>
      </header>
      <div
        class="StyledBox-sc-13pk1d4-0 jwyKWs"
      >
        <div
          class="StyledBox-sc-13pk1d4-0 hUSFEE StyledTabs-sc-a4fwxl-2 uzbFk"
          role="tablist"
        >
          <div
            class="StyledBox-sc-13pk1d4-0 jKAhip"
          >
            <div
              class="StyledBox-sc-13pk1d4-0 exoCoy StyledTabs__StyledTabsHeader-sc-a4fwxl-0 cUuna-D"
            >
              <button
                aria-expanded="true"
                aria-selected="true"
                class="StyledButton-sc-323bzc-0 hIPuDF"
                role="tab"
                type="button"
              >
                <div
                  class="StyledBox-sc-13pk1d4-0 jZDVC StyledTab-sc-1nnwnsb-0 jvjGeN"
                >
                  <span
                    class="StyledText-sc-1sadyjn-0 bouNPv"
                  >
                    githubrepo.tabs.0.title
                  </span>
                </div>
              </button>
              <button
                aria-expanded="false"
                aria-selected="false"
                class="StyledButton-sc-323bzc-0 hIPuDF"
                role="tab"
                type="button"
              >
                <div
                  class="StyledBox-sc-13pk1d4-0 etEeAW StyledTab-sc-1nnwnsb-0 jvjGeN"
                >
                  <span
                    class="StyledText-sc-1sadyjn-0 hnjXtp"
                  >
                    githubrepo.tabs.1.title (2)
                  </span>
                </div>
              </button>
            </div>
          </div>
          <div
            aria-label="githubrepo.tabs.0.title Tab Contents"
            class="StyledTabs__StyledTabPanel-sc-a4fwxl-1 fvSfiA"
            role="tabpanel"
          >
            <div
              class="StyledBox-sc-13pk1d4-0 jxnaYp"
            >
              <span
                class="StyledText-sc-1sadyjn-0 jbJJSg"
              >
                test Description
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`tests githubrepo card component tests render componen ==> fullsetTest no description 1`] = `
<div>
  <div
    class="StyledGrommet-sc-19lkkz7-0 ioQEen"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 hECOWc"
    >
      <header
        class="StyledBox-sc-13pk1d4-0 cUsmfa"
      >
        <div
          class="StyledBox-sc-13pk1d4-0 hyJwyv"
        >
          <div
            class="StyledBox-sc-13pk1d4-0 euWtCA"
          >
            <svg
              aria-label="Book"
              class="StyledIcon-sc-ofa7kd-0 jQvEgu"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 1v10l3-2 3 2V1M5.5 18a2.5 2.5 0 1 0 0 5H22M3 20.5v-17A2.5 2.5 0 0 1 5.5 1H21v17.007H5.492M20.5 18a2.5 2.5 0 1 0 0 5"
                fill="none"
                stroke="#000"
                stroke-width="2"
              />
            </svg>
            <div
              class="StyledBox__StyledBoxGap-sc-13pk1d4-1 ilMzAg"
            />
            <a
              class="StyledAnchor-sc-1rp7lwl-0 dSRcaq"
              href="owner html_url"
              target="_blank"
            >
              login
            </a>
            <div
              class="StyledBox__StyledBoxGap-sc-13pk1d4-1 ilMzAg"
            />
            <span
              class="StyledText-sc-1sadyjn-0 jbJJSg"
            >
              /
            </span>
            <div
              class="StyledBox__StyledBoxGap-sc-13pk1d4-1 ilMzAg"
            />
            <a
              class="StyledAnchor-sc-1rp7lwl-0 dSRcaq"
              href="repo html_url"
              target="_blank"
            >
              repo name
            </a>
          </div>
          <div
            class="StyledBox-sc-13pk1d4-0 byddCk"
          >
            <button
              class="StyledButton-sc-323bzc-0 bXHvkf"
              data-testid="bookmarkbutton"
              title="githubrepo.btnAddToFavoriteToolTip"
              type="button"
            >
              <svg
                aria-label="Bookmark"
                class="StyledIcon-sc-ofa7kd-0 iUTzQu"
                viewBox="0 0 24 24"
              >
                <path
                  d="M5 1v21l7-5 7 5V1z"
                  fill="none"
                  stroke="#000"
                  stroke-width="2"
                />
              </svg>
            </button>
            <div
              class="StyledBox__StyledBoxGap-sc-13pk1d4-1 drdMXk"
            />
            <div
              class="StyledBox-sc-13pk1d4-0 dEmnTc"
            >
              <svg
                aria-label="Star"
                class="StyledIcon-sc-ofa7kd-0 jQvEgu"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 16.667 5 22l3-8-6-4.5h7.5L12 2l2.5 7.5H22L16 14l3 8z"
                  fill="#FFC95E"
                  fill-rule="evenodd"
                />
              </svg>
              <div
                class="StyledBox__StyledBoxGap-sc-13pk1d4-1 cuTpwc"
              />
              <span
                class="StyledText-sc-1sadyjn-0 jbJJSg"
              >
                160
              </span>
            </div>
          </div>
        </div>
      </header>
      <div
        class="StyledBox-sc-13pk1d4-0 jwyKWs"
      >
        <div
          class="StyledBox-sc-13pk1d4-0 hUSFEE StyledTabs-sc-a4fwxl-2 uzbFk"
          role="tablist"
        >
          <div
            class="StyledBox-sc-13pk1d4-0 jKAhip"
          >
            <div
              class="StyledBox-sc-13pk1d4-0 exoCoy StyledTabs__StyledTabsHeader-sc-a4fwxl-0 cUuna-D"
            >
              <button
                aria-expanded="true"
                aria-selected="true"
                class="StyledButton-sc-323bzc-0 hIPuDF"
                role="tab"
                type="button"
              >
                <div
                  class="StyledBox-sc-13pk1d4-0 jZDVC StyledTab-sc-1nnwnsb-0 jvjGeN"
                >
                  <span
                    class="StyledText-sc-1sadyjn-0 bouNPv"
                  >
                    githubrepo.tabs.0.title
                  </span>
                </div>
              </button>
              <button
                aria-expanded="false"
                aria-selected="false"
                class="StyledButton-sc-323bzc-0 hIPuDF"
                role="tab"
                type="button"
              >
                <div
                  class="StyledBox-sc-13pk1d4-0 etEeAW StyledTab-sc-1nnwnsb-0 jvjGeN"
                >
                  <span
                    class="StyledText-sc-1sadyjn-0 hnjXtp"
                  >
                    githubrepo.tabs.1.title (2)
                  </span>
                </div>
              </button>
            </div>
          </div>
          <div
            aria-label="githubrepo.tabs.0.title Tab Contents"
            class="StyledTabs__StyledTabPanel-sc-a4fwxl-1 fvSfiA"
            role="tabpanel"
          >
            <div
              class="StyledBox-sc-13pk1d4-0 jxnaYp"
            >
              <span
                class="StyledText-sc-1sadyjn-0 jbJJSg"
              >
                githubrepo.tabs.0.noDescription
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`tests githubrepo card component tests render componen ==> fullsetTest not bookmarked 1`] = `
<div>
  <div
    class="StyledGrommet-sc-19lkkz7-0 ioQEen"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 hECOWc"
    >
      <header
        class="StyledBox-sc-13pk1d4-0 cUsmfa"
      >
        <div
          class="StyledBox-sc-13pk1d4-0 hyJwyv"
        >
          <div
            class="StyledBox-sc-13pk1d4-0 euWtCA"
          >
            <svg
              aria-label="Book"
              class="StyledIcon-sc-ofa7kd-0 jQvEgu"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 1v10l3-2 3 2V1M5.5 18a2.5 2.5 0 1 0 0 5H22M3 20.5v-17A2.5 2.5 0 0 1 5.5 1H21v17.007H5.492M20.5 18a2.5 2.5 0 1 0 0 5"
                fill="none"
                stroke="#000"
                stroke-width="2"
              />
            </svg>
            <div
              class="StyledBox__StyledBoxGap-sc-13pk1d4-1 ilMzAg"
            />
            <a
              class="StyledAnchor-sc-1rp7lwl-0 dSRcaq"
              href="owner html_url"
              target="_blank"
            >
              login
            </a>
            <div
              class="StyledBox__StyledBoxGap-sc-13pk1d4-1 ilMzAg"
            />
            <span
              class="StyledText-sc-1sadyjn-0 jbJJSg"
            >
              /
            </span>
            <div
              class="StyledBox__StyledBoxGap-sc-13pk1d4-1 ilMzAg"
            />
            <a
              class="StyledAnchor-sc-1rp7lwl-0 dSRcaq"
              href="repo html_url"
              target="_blank"
            >
              repo name
            </a>
          </div>
          <div
            class="StyledBox-sc-13pk1d4-0 byddCk"
          >
            <button
              class="StyledButton-sc-323bzc-0 bXHvkf"
              data-testid="bookmarkbutton"
              title="githubrepo.btnAddToFavoriteToolTip"
              type="button"
            >
              <svg
                aria-label="Bookmark"
                class="StyledIcon-sc-ofa7kd-0 jQvEgu"
                viewBox="0 0 24 24"
              >
                <path
                  d="M5 1v21l7-5 7 5V1z"
                  fill="none"
                  stroke="#000"
                  stroke-width="2"
                />
              </svg>
            </button>
            <div
              class="StyledBox__StyledBoxGap-sc-13pk1d4-1 drdMXk"
            />
            <div
              class="StyledBox-sc-13pk1d4-0 dEmnTc"
            >
              <svg
                aria-label="Star"
                class="StyledIcon-sc-ofa7kd-0 jQvEgu"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 16.667 5 22l3-8-6-4.5h7.5L12 2l2.5 7.5H22L16 14l3 8z"
                  fill="#FFC95E"
                  fill-rule="evenodd"
                />
              </svg>
              <div
                class="StyledBox__StyledBoxGap-sc-13pk1d4-1 cuTpwc"
              />
              <span
                class="StyledText-sc-1sadyjn-0 jbJJSg"
              >
                160
              </span>
            </div>
          </div>
        </div>
      </header>
      <div
        class="StyledBox-sc-13pk1d4-0 jwyKWs"
      >
        <div
          class="StyledBox-sc-13pk1d4-0 hUSFEE StyledTabs-sc-a4fwxl-2 uzbFk"
          role="tablist"
        >
          <div
            class="StyledBox-sc-13pk1d4-0 jKAhip"
          >
            <div
              class="StyledBox-sc-13pk1d4-0 exoCoy StyledTabs__StyledTabsHeader-sc-a4fwxl-0 cUuna-D"
            >
              <button
                aria-expanded="true"
                aria-selected="true"
                class="StyledButton-sc-323bzc-0 hIPuDF"
                role="tab"
                type="button"
              >
                <div
                  class="StyledBox-sc-13pk1d4-0 jZDVC StyledTab-sc-1nnwnsb-0 jvjGeN"
                >
                  <span
                    class="StyledText-sc-1sadyjn-0 bouNPv"
                  >
                    githubrepo.tabs.0.title
                  </span>
                </div>
              </button>
              <button
                aria-expanded="false"
                aria-selected="false"
                class="StyledButton-sc-323bzc-0 hIPuDF"
                role="tab"
                type="button"
              >
                <div
                  class="StyledBox-sc-13pk1d4-0 etEeAW StyledTab-sc-1nnwnsb-0 jvjGeN"
                >
                  <span
                    class="StyledText-sc-1sadyjn-0 hnjXtp"
                  >
                    githubrepo.tabs.1.title (2)
                  </span>
                </div>
              </button>
            </div>
          </div>
          <div
            aria-label="githubrepo.tabs.0.title Tab Contents"
            class="StyledTabs__StyledTabPanel-sc-a4fwxl-1 fvSfiA"
            role="tabpanel"
          >
            <div
              class="StyledBox-sc-13pk1d4-0 jxnaYp"
            >
              <span
                class="StyledText-sc-1sadyjn-0 jbJJSg"
              >
                test Description
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
